File: backend/app/core/evaluate.py
"""
Head-to-head evaluation: exact expected value, Monte-Carlo simulation, and the
statistics needed to say whether a measured edge is real.

Three ideas drive the design.

Seat matters, so both seats get played
--------------------------------------
Kuhn poker is not symmetric: the first player is worth -1/18 chips per hand at
equilibrium. Comparing a strategy in seat 0 against a baseline in seat 1 would
therefore measure the seat as much as the strategy. Every result here is
*duplicate-scored*: each pairing is played from both seats and the two figures
averaged, which cancels the positional term exactly.

Exact beats sampled, when exact is available
--------------------------------------------
Both a CFR solution and a rule-based baseline are fixed strategies, so their
head-to-head expected value can be computed by enumerating all six deals and
every branch of the tree. That is a closed-form number with no confidence
interval attached. The Monte-Carlo path exists to *validate* the exact
computation and to produce realistic hand-by-hand distributions, not because
the estimate is needed.

An estimate without an interval is not a result
-----------------------------------------------
Simulated win rates come back with a standard error, a 95% confidence interval
and a p-value against the null hypothesis of a zero edge. Poker results have
enormous variance relative to their means: in this game a single hand swings
+-2 chips while a strong edge is ~0.2 chips per hand, so a few hundred hands
can easily show the wrong sign. The interval is what separates a measured edge
from a lucky one.
"""

from __future__ import annotations

import math
import random
from typing import Dict, List, Optional, Sequence, Tuple

from .best_response import best_response, best_response_strategy, exploitability
from .kuhn import (
    ACTIONS, DEAL_PROB, DEALS,
    current_player, info_set_key, is_terminal, terminal_utility,
)

Strategy = Dict[str, List[float]]
# Either one strategy used in both seats, or a (player0, player1) pair. The
# exploitative agent needs the pair form, because the best response to an
# opponent depends on which seat you are sitting in.
SeatedStrategy = object


def _seat(strategy, seat: int) -> Strategy:
    """Resolve a strategy-or-pair down to the strategy for one seat."""
    if isinstance(strategy, tuple):
        return strategy[seat]
    return strategy


def exploitative_agent(opponent: Strategy) -> Tuple[Strategy, Strategy]:
    """
    The maximally exploitative counter-strategy to `opponent`, one per seat.

    Contrast with the equilibrium: this agent wins the most possible against
    this specific opponent, and would be crushed by an opponent who noticed.
    """
    return best_response_strategy(opponent, 0), best_response_strategy(opponent, 1)


# --------------------------------------------------------------------------
# Exact evaluation
# --------------------------------------------------------------------------

def exact_ev(p0_strategy: Strategy, p1_strategy: Strategy) -> float:
    """
    Exact expected value to Player 0 when the two seats use different
    strategies. Enumerates every deal and every branch; no sampling.
    """
    def walk(deal, history: str) -> float:
        if is_terminal(history):
            return terminal_utility(history, deal)
        player = current_player(history)
        source = p0_strategy if player == 0 else p1_strategy
        probs = source[info_set_key(deal[player], history)]
        total = 0.0
        for i, action in enumerate(ACTIONS):
            if probs[i] > 0.0:
                total += probs[i] * walk(deal, history + action)
        return total

    return sum(DEAL_PROB * walk(deal, "") for deal in DEALS)


def duplicate_ev(strategy, opponent: Strategy) -> Dict[str, float]:
    """
    Seat-balanced expected value to `strategy`, in chips per hand.

    Playing both seats removes the structural -1/18 first-player disadvantage,
    so the number that comes back reflects strategy quality alone.
    """
    as_p0 = exact_ev(_seat(strategy, 0), opponent)        # value to us, seated first
    as_p1 = -exact_ev(opponent, _seat(strategy, 1))       # value to us, seated second
    return {
        "ev_as_player0": as_p0,
        "ev_as_player1": as_p1,
        "ev_per_hand": (as_p0 + as_p1) / 2.0,
        "milli_antes_per_hand": (as_p0 + as_p1) / 2.0 * 1000.0,
    }


def exploitative_gap(strategy, opponent: Strategy) -> Dict[str, float]:
    """
    How much expected value equilibrium play leaves on the table against a
    specific flawed opponent.

    This is the most misunderstood property of a Nash strategy, and worth
    stating plainly: an equilibrium does not try to win the maximum. It
    guarantees it cannot *lose* more than the game value, whoever it faces. A
    best response to this particular opponent earns strictly more, but is
    itself wide open to a counter-adjustment. The gap below is the price of
    that safety.
    """
    nash_side = duplicate_ev(strategy, opponent)["ev_per_hand"]
    br_as_p0, _ = best_response(opponent, 0)
    br_as_p1, _ = best_response(opponent, 1)
    max_exploit = (br_as_p0 + br_as_p1) / 2.0
    return {
        "equilibrium_ev": nash_side,
        "max_exploit_ev": max_exploit,
        "gap": max_exploit - nash_side,
    }


# --------------------------------------------------------------------------
# Monte-Carlo simulation
# --------------------------------------------------------------------------

def _play_hand(deal: Tuple[int, int], p0_strategy: Strategy,
               p1_strategy: Strategy, rng: random.Random) -> Tuple[float, str]:
    """Play one hand, sampling each decision from the acting player's strategy."""
    history = ""
    while not is_terminal(history):
        player = current_player(history)
        source = p0_strategy if player == 0 else p1_strategy
        probs = source[info_set_key(deal[player], history)]
        history += ACTIONS[0] if rng.random() < probs[0] else ACTIONS[1]
    return terminal_utility(history, deal), history


def simulate(strategy, opponent: Strategy, hands: int = 100_000,
             seed: Optional[int] = 42) -> Dict[str, object]:
    """
    Play `hands` hands, alternating seats, and report the result with its
    uncertainty.

    Seats alternate every hand so the sample is seat-balanced by construction,
    matching the exact duplicate figure it is checked against.
    """
    if hands <= 0:
        raise ValueError("hands must be positive")

    rng = random.Random(seed)
    total = 0.0
    total_sq = 0.0
    wins = losses = 0
    showdowns = folds = 0
    by_seat = {0: [0.0, 0], 1: [0.0, 0]}

    for i in range(hands):
        we_are_p0 = (i % 2 == 0)
        deal = DEALS[rng.randrange(len(DEALS))]
        if we_are_p0:
            payoff, history = _play_hand(deal, _seat(strategy, 0), opponent, rng)
        else:
            payoff, history = _play_hand(deal, opponent, _seat(strategy, 1), rng)
            payoff = -payoff  # flip to our perspective

        total += payoff
        total_sq += payoff * payoff
        if payoff > 0:
            wins += 1
        elif payoff < 0:
            losses += 1
        if history in ("pp", "bb", "pbb"):
            showdowns += 1
        else:
            folds += 1
        seat = 0 if we_are_p0 else 1
        by_seat[seat][0] += payoff
        by_seat[seat][1] += 1

    mean = total / hands
    # Population variance of the per-hand payoff, then the standard error.
    variance = max(0.0, total_sq / hands - mean * mean)
    stdev = math.sqrt(variance)
    sem = stdev / math.sqrt(hands) if hands > 1 else float("inf")

    # Null hypothesis: the true per-hand edge is zero. With n in the tens of
    # thousands the t-distribution is indistinguishable from the normal, so a
    # z-test is used and reported as such.
    z = mean / sem if sem > 0 else 0.0
    p_value = 2.0 * (1.0 - _normal_cdf(abs(z)))
    half_width = 1.959963985 * sem  # 95%

    exact = duplicate_ev(strategy, opponent)["ev_per_hand"]

    return {
        "hands": hands,
        "seed": seed,
        "ev_per_hand": mean,
        "stdev": stdev,
        "standard_error": sem,
        "ci95_low": mean - half_width,
        "ci95_high": mean + half_width,
        "z_score": z,
        "p_value": p_value,
        "significant_at_95": p_value < 0.05,
        "win_rate": wins / hands,
        "loss_rate": losses / hands,
        "wins": wins,
        "losses": losses,
        "showdown_rate": showdowns / hands,
        "fold_rate": folds / hands,
        "ev_as_player0": by_seat[0][0] / by_seat[0][1] if by_seat[0][1] else 0.0,
        "ev_as_player1": by_seat[1][0] / by_seat[1][1] if by_seat[1][1] else 0.0,
        "exact_ev_per_hand": exact,
        # The validation that matters: does the sampled mean land inside its own
        # interval around the analytically known answer?
        "exact_within_ci95": (mean - half_width) <= exact <= (mean + half_width),
        "absolute_error_vs_exact": abs(mean - exact),
        "standard_errors_from_exact": abs(mean - exact) / sem if sem > 0 else 0.0,
    }


def _normal_cdf(x: float) -> float:
    """Standard normal CDF via the error function."""
    return 0.5 * (1.0 + math.erf(x / math.sqrt(2.0)))


def required_hands(edge: float, stdev: float = 1.0, z: float = 1.959963985) -> int:
    """
    Hands needed to resolve an edge of `edge` chips/hand at 95% confidence.

    Included because it is the number that surprises people: the sample size
    scales with the *square* of the inverse edge, so halving the edge you want
    to detect quadruples the hands you need.
    """
    if edge <= 0:
        raise ValueError("edge must be positive")
    return int(math.ceil((z * stdev / edge) ** 2))


# --------------------------------------------------------------------------
# Tournament
# --------------------------------------------------------------------------

def evaluate_against_baselines(strategy,
                               baselines: Dict[str, Strategy],
                               hands: int = 100_000,
                               seed: Optional[int] = 42,
                               simulate_too: bool = True) -> List[Dict[str, object]]:
    """Play the solved strategy against every baseline and collect the results."""
    rows: List[Dict[str, object]] = []
    for name, opponent in sorted(baselines.items()):
        row: Dict[str, object] = {"opponent": name}
        row.update(duplicate_ev(strategy, opponent))
        row.update(exploitative_gap(strategy, opponent))
        row["opponent_exploitability"] = exploitability(opponent)
        if simulate_too:
            row["simulation"] = simulate(strategy, opponent, hands=hands, seed=seed)
        rows.append(row)
    rows.sort(key=lambda r: -float(r["ev_per_hand"]))
    return rows


__all__ = [
    "exact_ev", "duplicate_ev", "exploitative_gap", "simulate",
    "required_hands", "evaluate_against_baselines", "exploitative_agent",
]

File: frontend/lib/solver/evaluate.ts
/**
 * Head-to-head evaluation: exact expected value, Monte-Carlo simulation, and
 * the statistics needed to say whether a measured edge is real. Port of
 * `backend/app/core/evaluate.py`.
 *
 * Seat matters, so both seats get played. Kuhn poker is not symmetric - the
 * first player is worth -1/18 chips per hand at equilibrium - so every result
 * here is *duplicate-scored*: each pairing is played from both seats and
 * averaged, which cancels the positional term exactly.
 *
 * Exact beats sampled when exact is available. Both a CFR solution and a
 * rule-based baseline are fixed strategies, so their head-to-head EV can be
 * enumerated. The Monte-Carlo path exists to *validate* that number and to
 * produce realistic hand-by-hand variance, not because an estimate is needed.
 *
 * An estimate without an interval is not a result. In this game a single hand
 * swings +-2 chips while a strong edge is ~0.2 chips/hand, so a few hundred
 * hands can easily show the wrong sign.
 */

import { bestResponse, bestResponseStrategy, exploitability } from "./bestResponse";
import {
  ACTIONS, DEALS, DEAL_PROB,
  currentPlayer, infoSetKey, isTerminal, terminalUtility,
  type Deal, type Strategy,
} from "./kuhn";

/** One strategy used in both seats, or a [player0, player1] pair. The
 *  exploitative agent needs the pair form, because the best response to an
 *  opponent depends on which seat you sit in. */
export type Seated = Strategy | [Strategy, Strategy];

function seat(strategy: Seated, index: number): Strategy {
  return Array.isArray(strategy) ? strategy[index] : strategy;
}

/**
 * The maximally exploitative counter-strategy to `opponent`, one per seat.
 * Contrast with the equilibrium: this wins the most possible against this
 * specific opponent, and would be crushed by an opponent who noticed.
 */
export function exploitativeAgent(opponent: Strategy): [Strategy, Strategy] {
  return [bestResponseStrategy(opponent, 0), bestResponseStrategy(opponent, 1)];
}

/** Exact EV to Player 0 when the two seats use different strategies. */
export function exactEv(p0: Strategy, p1: Strategy): number {
  const walk = (deal: Deal, history: string): number => {
    if (isTerminal(history)) return terminalUtility(history, deal);
    const player = currentPlayer(history);
    const probs = (player === 0 ? p0 : p1)[infoSetKey(deal[player], history)];
    let total = 0;
    for (let i = 0; i < ACTIONS.length; i++) {
      if (probs[i] > 0) total += probs[i] * walk(deal, history + ACTIONS[i]);
    }
    return total;
  };
  let total = 0;
  for (const deal of DEALS) total += DEAL_PROB * walk(deal, "");
  return total;
}

export interface DuplicateEv {
  ev_as_player0: number;
  ev_as_player1: number;
  ev_per_hand: number;
  milli_antes_per_hand: number;
}

/**
 * Seat-balanced EV to `strategy`, in chips per hand. Playing both seats removes
 * the structural -1/18 first-player disadvantage, so the number reflects
 * strategy quality alone.
 */
export function duplicateEv(strategy: Seated, opponent: Strategy): DuplicateEv {
  const asP0 = exactEv(seat(strategy, 0), opponent);
  const asP1 = -exactEv(opponent, seat(strategy, 1));
  const mean = (asP0 + asP1) / 2;
  return {
    ev_as_player0: asP0,
    ev_as_player1: asP1,
    ev_per_hand: mean,
    milli_antes_per_hand: mean * 1000,
  };
}

/**
 * How much EV equilibrium play leaves on the table against a specific flawed
 * opponent.
 *
 * The most misunderstood property of a Nash strategy, worth stating plainly: an
 * equilibrium does not try to win the maximum. It guarantees it cannot *lose*
 * more than the game value, whoever it faces. A best response to this
 * particular opponent earns strictly more, but is itself wide open to a
 * counter-adjustment. The gap is the price of that safety.
 */
export function exploitativeGap(strategy: Seated, opponent: Strategy): {
  equilibrium_ev: number; max_exploit_ev: number; gap: number;
} {
  const nashSide = duplicateEv(strategy, opponent).ev_per_hand;
  const maxExploit =
    (bestResponse(opponent, 0).value + bestResponse(opponent, 1).value) / 2;
  return {
    equilibrium_ev: nashSide,
    max_exploit_ev: maxExploit,
    gap: maxExploit - nashSide,
  };
}

/**
 * mulberry32: a small, fast, seedable PRNG.
 *
 * The Python reference uses the Mersenne Twister, so simulated streams differ
 * between the two implementations. That is expected and harmless - the
 * simulation exists to validate the *exact* enumerated EV, and that exact
 * number is identical in both. What matters here is only that runs are
 * reproducible, which a seeded generator gives.
 */
function mulberry32(seed: number): () => number {
  let a = seed >>> 0;
  return () => {
    a = (a + 0x6d2b79f5) >>> 0;
    let t = a;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  };
}

/** Standard normal CDF, via an Abramowitz-Stegun approximation to erf. */
function normalCdf(x: number): number {
  const t = 1 / (1 + 0.3275911 * Math.abs(x) / Math.SQRT2);
  const y = 1 - ((((1.061405429 * t - 1.453152027) * t + 1.421413741) * t
    - 0.284496736) * t + 0.254829592) * t * Math.exp(-(x * x) / 2);
  return x >= 0 ? 0.5 * (1 + y) : 0.5 * (1 - y);
}

export interface Simulation {
  hands: number;
  ev_per_hand: number;
  stdev: number;
  standard_error: number;
  ci95_low: number;
  ci95_high: number;
  p_value: number;
  significant_at_95: boolean;
  win_rate: number;
  showdown_rate: number;
  fold_rate: number;
  exact_ev_per_hand: number;
  exact_within_ci95: boolean;
  standard_errors_from_exact: number;
}

function playHand(deal: Deal, p0: Strategy, p1: Strategy, rng: () => number): [number, string] {
  let history = "";
  while (!isTerminal(history)) {
    const player = currentPlayer(history);
    const probs = (player === 0 ? p0 : p1)[infoSetKey(deal[player], history)];
    history += rng() < probs[0] ? ACTIONS[0] : ACTIONS[1];
  }
  return [terminalUtility(history, deal), history];
}

/**
 * Play `hands` hands, alternating seats every hand so the sample is
 * seat-balanced by construction, and report the result with its uncertainty.
 */
export function simulate(strategy: Seated, opponent: Strategy,
                         hands = 100_000, seed = 42): Simulation {
  if (hands <= 0) throw new Error("hands must be positive");
  const rng = mulberry32(seed);
  let total = 0, totalSq = 0, wins = 0, showdowns = 0;

  for (let i = 0; i < hands; i++) {
    const deal = DEALS[Math.floor(rng() * DEALS.length)];
    let payoff: number, history: string;
    if (i % 2 === 0) {
      [payoff, history] = playHand(deal, seat(strategy, 0), opponent, rng);
    } else {
      [payoff, history] = playHand(deal, opponent, seat(strategy, 1), rng);
      payoff = -payoff; // flip to our perspective
    }
    total += payoff;
    totalSq += payoff * payoff;
    if (payoff > 0) wins++;
    if (history === "pp" || history === "bb" || history === "pbb") showdowns++;
  }

  const mean = total / hands;
  const variance = Math.max(0, totalSq / hands - mean * mean);
  const stdev = Math.sqrt(variance);
  const sem = hands > 1 ? stdev / Math.sqrt(hands) : Infinity;
  // Null hypothesis: the true per-hand edge is zero. With n in the tens of
  // thousands the t-distribution is indistinguishable from the normal.
  const z = sem > 0 ? mean / sem : 0;
  const pValue = 2 * (1 - normalCdf(Math.abs(z)));
  const halfWidth = 1.959963985 * sem;
  const exact = duplicateEv(strategy, opponent).ev_per_hand;

  return {
    hands,
    ev_per_hand: mean,
    stdev,
    standard_error: sem,
    ci95_low: mean - halfWidth,
    ci95_high: mean + halfWidth,
    p_value: pValue,
    significant_at_95: pValue < 0.05,
    win_rate: wins / hands,
    showdown_rate: showdowns / hands,
    fold_rate: (hands - showdowns) / hands,
    exact_ev_per_hand: exact,
    // The validation that matters: does the sampled mean land inside its own
    // interval around the analytically known answer?
    exact_within_ci95: mean - halfWidth <= exact && exact <= mean + halfWidth,
    standard_errors_from_exact: sem > 0 ? Math.abs(mean - exact) / sem : 0,
  };
}

/**
 * Hands needed to resolve an edge at 95% confidence. Included because the
 * number surprises people: sample size scales with the *square* of the inverse
 * edge, so halving the edge you want to detect quadruples the hands you need.
 */
export function requiredHands(edge: number, stdev = 1, z = 1.959963985): number {
  if (edge <= 0) throw new Error("edge must be positive");
  return Math.ceil(((z * stdev) / edge) ** 2);
}

export { exploitability };
